### ISSUE TITLE
Change Request: Allow rules to specify the languages/dialects they work on

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#19462]
### ESLint version

HEAD

### What problem do you want to solve?

Currently, a rule has no way to indicate if it works on any given language. This information would be helpful for two reasons:

1. Documentation purposes - being able to see which JS rules have been updated to support TS syntax.
2. Runtime purposes - automatically turning off rules that don't apply to the given language (for instance, turning off JS rules when linting CSS files)

### What do you think is the correct solution?

In the 2025-02-20 TSC meeting, we discussed adding two properties to `meta`:

1. `meta.language` - a string, such as "javascript", "css"
2. `meta.dialects` - an array of strings, such as `["javascript", "typescript"]`

@JoshuaKGoldberg [suggested](https://github.com/eslint/eslint/pull/19431/files/de59b04450d780f0e5c4f8826c318f8d263438f0..b7b964a3d87fa0044ffab51cbe3fe69bf514b6b9#r1968176879) merging both properties into a single object instead.

It's probably worth putting together an RFC that takes into account the two use cases (documentation and runtime) as well as finalizing what the `meta` info should look like.

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by fasttime** [comment]
I think it's a good idea to have a single object grouping the two properties (currently named `language` and `dialects`) as suggested by @JoshuaKGoldberg. We'll want `dialects` to be an optional property inside the new object.

Rules that don't match the language hint to a problem with the config. I can think of several options to handle that situation:
* Ignore the rule(s)
* Ignore the rule(s) and emit a warning
* Throw an error

Rules without a `language` property will be still allowed to run on every language, meaning that the new behavior with rules that have a `language` property associated should not be a breaking change if it's implemented from the beginning.

By contrast, letting a rule run with an unsupported AST or language-related data seems to open the door for undefined behavior, so I'd suggest removing that option.

**Comment by mdjermanovic** [comment]
I support this proposal. I'm not sure if ESLlint should automatically turn off rules that don't apply to the given language or rather throw a descriptive error because end users' expectations might be that some rules (e.g., `no-irregular-whitespace`) work for any language, in which case silently ignoring enabled rules would be surprising. We can discuss that in the RFC.

**Comment by fasttime** [comment]
Not stale, this is awaiting feedback @eslint/eslint-team.

**Comment by nzakas** [comment]
Thinking through this, we may actually have two different use cases here:

1. To determine if a rule will work for a given language. This actually relates more to the plugin language than it does the formal name of the language (so, the `json` language in the `json` plugin vs. "JSON"). A rule may say it supports JSON but if someone is using a different language implementation that outputs a different AST, then this data isn't useful.
2. To document if a rule works with a particular dialect of a language, the canonical example being if a rule is known to work with JavaScript and TypeScript. In this case, we might say that a rule supports the `js/js` language and due to the use of a custom parser, it can also support TypeScript. This has no functional effect on the rule, it's just basically documentation.

So what we might want to do is two different things:

```js
export default {
    meta: {

        // indicate which plugin languages are required
        languages: ["markdown/commonmark", "markdown/gfm"]

        // indicate which language dialects the rule logically supports
        docs: {
            dialects: ["Markdown", "GFM", "MDX"]
        }
}
```

**Comment by nzakas** [comment]
ping @eslint/eslint-tsc

**Comment by mdjermanovic** [comment]
> 1. To determine if a rule will work for a given language. This actually relates more to the plugin language than it does the formal name of the language (so, the `json` language in the `json` plugin vs. "JSON"). A rule may say it supports JSON but if someone is using a different language implementation that outputs a different AST, then this data isn't useful.

> ```
> export default {
>     meta: {
> 
>         // indicate which plugin languages are required
>         languages: ["markdown/commonmark", "markdown/gfm"]
> ```

How would we match "markdown" with `@eslint/markdown` specifically, and ensure it doesn't match with different language implementations that might appear in the future but output a different AST?

**Comment by fasttime** [comment]
> So what we might want to do is two different things:
> 
> ```js
> export default {
>     meta: {
> 
>         // indicate which plugin languages are required
>         languages: ["markdown/commonmark", "markdown/gfm"]
> 
>         // indicate which language dialects the rule logically supports
>         docs: {
>             dialects: ["Markdown", "GFM", "MDX"]
>         }
> }
> ```

I'm fine with this approach in general. Two things I'd like to suggest if we decide to go this way:

* It would be useful to have a match-all syntax for languages that share a common base:
  ```js
  // Works with any languages of the `markdown` family.
  languages: ["markdown/*"],
  ```

  or for rules that support any language like `no-restricted-syntax`:

  ```js
  languages: ["*"],
  ```

* Dialects should be required to use the standard capitalization for language names and to avoid abbreviations, i.e. `TypeScript` instead of `typescript`, `TS`, etc.

**Comment by fasttime** [comment]
@eslint/eslint-tsc How do we want to procede with this change request?

**Comment by mdjermanovic** [comment]
> My idea was that we'd look up from the rule like this:
> 
> 1. Find the plugin in the config named `markdown`. If one doesn't exist, then search the registered plugins for one with a `namespace` of `"markdown"`.
> 2. Then see if there's a language called `markdown`.
> 3. If we don't find the language, then disable the rule.
> 
> This is a highly cacheable query so we'd really only do the lookup one time per language per config.

The described matching sounds good to me.

I'm only not sure if we should disable the rule or rather throw a descriptive error. Silently disabling things that might have been expected to work seems dangerous because users wouldn't notice it. Now that we have `defineConfig()` with `extends`, it's much easier to apply rules to correct files.

**Comment by nzakas** [comment]
Marking as accepted. This will need an RFC to discuss all of the details.

Does anyone want to take this? If not, I can.

**Comment by nzakas** [comment]
Working on this.

**Comment by mdjermanovic** [comment]
@nzakas are there any remaining tasks here or did we just forgot to link this issue on https://github.com/eslint/eslint/pull/20571 so it wasn't autoclosed?

**Comment by nzakas** [comment]
Looks like I just forgot to link it. All the work is done.
